vulkano/src/shader/mod.rs

```diff
@@ -16,6 +16,86 @@
+//!
+//! When buffers, push constants or other user-provided data are accessed in shaders,
+//! the shader expects the values to be laid out according to specific rules.
+//! As in Rust, the location in memory (or offset from the start of a buffer) of each value
+//! accessed by a shader must be a multiple of a certain number, which is called its *alignment*.
+//! Values used in shaders do not necessarily have the same alignment as their counterparts in
+//! Rust or C. You must therefore make sure that your data is placed at the locations the shader
+//! expects, or it will read the wrong data and produce nonsense.
+//!
+//! Three sets of [alignment rules] are supported by Vulkan, listed from least to most
+//! restrictive:
+//!
+//! - **Scalar alignment**, which is the same as the C alignment, expressed in Rust with the
+//!   [`#[repr(C)]`](https://doc.rust-lang.org/nomicon/other-reprs.html#reprc) attribute.
+//!   This is not used by default anywhere, but can be used for certain cases by enabling the
+//!   [`scalar_block_layout`] feature, and prefixing the block in GLSL with `layout(scalar)`
+//!   (this requires the [`GL_EXT_scalar_block_layout`] GLSL extension).
+//! - **Base alignment**, also known as **std430**.
+//!   This is used by default for all shader data except uniform buffers. It can also be used
+//!   for uniform buffers, by enabling the [`uniform_buffer_standard_layout`] feature, and
+//!   prefixing the block in GLSL with `layout(std430)`.
+//! - **Extended alignment**, also known as **std140**.
+//!   This is used by default for uniform buffers.
+//!
+//! In all three of these alignment rules, a primitive/scalar value with a size of N bytes has an
+//! alignment of N, meaning that it must have an offset in the buffer/memory that is a multiple of
+//! its size, like in C or Rust. So, for example, a `float` (like a Rust `f32`) has a size of 4 bytes, and
+//! an alignment of 4.
+//!
+//! The differences between the alignment rules are in how compound types (vectors, matrices,
+//! arrays and structs) are expected to be laid out. For a compound type with an element whose
+//! alignment is N, the scalar alignment considers the alignment of the compound type to be also N.
+//! However, the base and extended alignment are stricter:
+//!
+//! | GLSL type | Scalar         | Base           | Extended                                     |
+//! |-----------|----------------|----------------|----------------------------------------------|
+//! | primitive | N              | N              | N                                            |
+//! | `vec2`    | N              | N * 2          | N * 2                                        |
+//! | `vec3`    | N              | N * 4          | N * 4                                        |
+//! | `vec4`    | N              | N * 4          | N * 4                                        |
+//! | array     | N              | N              | N, rounded up to multiple of 16              |
```

Comment:
I think this is expressed more simply as:
```suggestion
//! | array     | N              | N              | max of N and 16                              |
```
The specification expresses things in a really convoluted way...

Reply:
That's not exactly the same though. If N were, for example, 24, then it would be rounded up to 32.

Reply:
Alignments must be a power of two though, which is why taking the maximum always works.

Reply:
Ah right, good point.